contracts/ReversibleICO.sol

```diff
@@ -15,169 +15,165 @@ import "./zeppelin/introspection/IERC1820Registry.sol";
+        address _TokenContractAddress,
         address _whitelistControllerAddress,
         address _projectWalletAddress,
-        uint256 _StartBlock,
-        uint256 _AllocationBlockCount,
-        uint256 _AllocationPrice,
+        uint256 _commitPhaseStartBlock,
+        uint256 _commitPhaseBlockCount,
+        uint256 _commitPhasePrice,
         uint8   _StageCount,
         uint256 _StageBlockCount,
         uint256 _StagePriceIncrease
     )
         public
         onlyDeployer
-        requireNotInitialized
+        notInitialized
     {
-        // addresses
-        TokenTrackerAddress = _TokenTrackerAddress;
+
+        // Assign address variables
+        TokenContractAddress = _TokenContractAddress;
         whitelistControllerAddress = _whitelistControllerAddress;
         projectWalletAddress = _projectWalletAddress;
 
-        // initialize ERC777 TokenTracker
-        TokenTracker = IERC777(TokenTrackerAddress);
-
-        // Allocation settings
-        StartBlock = _StartBlock;
-        AllocationBlockCount = _AllocationBlockCount;
-        AllocationEndBlock = StartBlock + AllocationBlockCount;
-        AllocationPrice = _AllocationPrice;
+        // Assign other variables
+        commitPhaseStartBlock = _commitPhaseStartBlock;
+        commitPhaseBlockCount = _commitPhaseBlockCount;
+        commitPhaseEndBlock = _commitPhaseStartBlock + _commitPhaseBlockCount;
+        commitPhasePrice = _commitPhasePrice;
 
         StageBlockCount = _StageBlockCount;
 
-        // first stage is allocation. Set it up.
-        ContractStage storage StageRecord = StageByNumber[ContractStageCount];
-        StageRecord.start_block = _StartBlock;
-        StageRecord.end_block = _StartBlock + _AllocationBlockCount;
-        StageRecord.token_price = _AllocationPrice;
-        ContractStageCount++;
 
-        uint256 lastStageBlockEnd = StageRecord.end_block;
+        // initialize ERC777 TokenContract
+        TokenContract = IERC777(TokenContractAddress);
+
+
+        // Setup stage 0: The commit phase.
+        Stage storage Stage0 = Stages[StageCount]; // StageCount = 0
+        Stage0.start_block = _commitPhaseStartBlock;
+        Stage0.end_block = _commitPhaseStartBlock + _commitPhaseBlockCount;
+        Stage0.token_price = _commitPhasePrice;
+
+        StageCount++; // StageCount = 1
+
+
+        // Setup stage 1 to n: The buy phase stages
+        uint256 lastStageBlockEnd = Stage0.end_block;
 
-        // calculate block ranges and set price for each period
         for(uint8 i = 1; i <= _StageCount; i++) {
 
-            StageRecord = StageByNumber[ContractStageCount];
-            StageRecord.start_block = lastStageBlockEnd + 1;
-            StageRecord.end_block = lastStageBlockEnd + _StageBlockCount + 1;
-            StageRecord.token_price = _AllocationPrice + ( _StagePriceIncrease * (i) );
-            ContractStageCount++;
+            Stage storage StageN = Stages[StageCount]; // StageCount = n
```

Comment:
Creating a new object here for every loop, does this make the execution a lot more expensive?

Reply:
Not really
not initialising a StageRecord does make it a lot more expensive

### maybe it can be improved but no idea how.
```
            StageRecord = StageByNumber[ContractStageCount];
            StageRecord.start_block = lastStageBlockEnd + 1;
            StageRecord.end_block = lastStageBlockEnd + _StageBlockCount + 1;
            StageRecord.token_price = _AllocationPrice + ( _StagePriceIncrease * (i) );
            ContractStageCount++;
```

### costly since VM needs to look up the stack item each time
```
            StageByNumber[ContractStageCount].start_block = lastStageBlockEnd + 1;
            StageByNumber[ContractStageCount].end_block = lastStageBlockEnd + _StageBlockCount + 1;
            StageByNumber[ContractStageCount].token_price = _AllocationPrice + ( _StagePriceIncrease * (i) );
            ContractStageCount++;
```

### same as ParticipantTotalsDetails method